modules/cnv-troubleshooting-incorrect-policy-config.adoc

```diff
@@ -0,0 +1,133 @@
+
+[id="cnv-troubleshooting-incorrect-policy-config_{context}"]
+= Troubleshooting: Incorrect NodeNetworkConfigurationPolicy configuration
+
+Applying a NodeNetworkConfigurationPolicy to the cluster is an easy way to apply changes to the node network configuration consistently across your entire cluster.
+However, this also means it is easy to apply an incorrect configuration.
+
+This troublshooting example investigates how to identify and correct the configuration of a failed network Policy.
+
+In this example, a Linux bridge Policy is applied to an example cluster that has 3 master nodes and 3 worker nodes. The Policy fails to be applied because it references an incorrect interface. The error can be found by investigating the available nmstate resources and corrected by updating the Policy. 
+
+.Procedure
+
+. Create a Policy and apply it to cluster. The following example creates a simple bridge on the `ens01` interface:
++
+[source,yaml]
+----
+apiVersion: nmstate.io/v1alpha1
+kind: NodeNetworkConfigurationPolicy
+metadata:
+  name: ens01-bridge-testfail
+spec:
+  desiredState:
+    interfaces:
+      - name: br1
+        description: Linux bridge with the wrong port 
+        type: linux-bridge
+        state: up
+        ipv4:
+          dhcp: true
+          enabled: true
+        bridge:
+          options:
+            stp:
+              enabled: false
+          port:
+            - name: ens01
+----
++
+----
+nodenetworkconfigurationpolicy.nmstate.io/ens01-bridge-testfail created
+----
+
+. It can take up to 20 seconds for the policy to apply, after which we can check the status of the Policy to ensure it has been successfully configured:
++
+----
+$ oc get nncp
+----
++
+The response shows that the Policy failed:
++
+----
+NAME                    STATUS
+ens01-bridge-testfail   FailedToConfigure
+----
+
+. The Policy status alone does not tell us if it failed on all nodes or a subset of nodes. Listing the Enactments we can see if the Policy was successful on any of the nodes. If the Policy failed for only a subset it suggests the problem is with specific node configuration; if the Policy failed on all nodes it suggest the problem is with the Policy.
++
+----
+$ oc get nnce
+----
++
+The response shows that the Policy failed on all nodes:
++
+----
+NAME                                                STATUS
+master-1.ens01-bridge-testfail         FailedToConfigure
+master-2.ens01-bridge-testfail         FailedToConfigure
+master-3.ens01-bridge-testfail         FailedToConfigure
+worker-1.ens01-bridge-testfail         FailedToConfigure
+worker-2.ens01-bridge-testfail         FailedToConfigure
+worker-3.ens01-bridge-testfail         FailedToConfigure
+----
+
+. Viewing one of the Enactments we can look at the traceback. To make the output simpler for this example, we'll pipe the output to grep and search for 'error' without case sensitivity:
++
+----
+$ oc get nnce worker-1.ens01-bridge-testfail -o yaml | grep -i 'error'
```

Comment:
There is an alternative. You tell me if it makes more sense.

`oc get nnce worker-1.ens01-bridge-testfail -o jsonpath='{.status.conditions[?(@.type=="Failing")].message}'`

It would show something like:


```
error reconciling NodeNetworkConfigurationPolicy at desired state apply: , failed to execute nmstatectl set --no-commit --timeout 480: 'exit status 1' '' '2020-04-09 10:26:45,681 root         DEBUG    Checkpoint
 /org/freedesktop/NetworkManager/Checkpoint/1 created for all devices: 480                                                                                                                                        
2020-04-09 10:26:45,681 root         DEBUG    Adding new interfaces: ['br1']
2020-04-09 10:26:45,683 root         DEBUG    Editing interfaces: []
2020-04-09 10:26:45,685 root         DEBUG    Executing NM action: func=add_connection_async
2020-04-09 10:26:45,764 root         DEBUG    Connection adding succeeded: dev=br1
2020-04-09 10:26:45,770 root         DEBUG    Executing NM action: func=safe_activate_async
2020-04-09 10:26:45,778 root         DEBUG    Connection activation initiated: dev=br1, con-state=<enum NM_ACTIVE_CONNECTION_STATE_ACTIVATING of type NM.ActiveConnectionState>
2020-04-09 10:26:45,830 root         DEBUG    Connection activation succeeded: dev=br1, con-state=<enum NM_ACTIVE_CONNECTION_STATE_ACTIVATING of type NM.ActiveConnectionState>, dev-state=<enum NM_DEVICE_STATE_IP
_CONFIG of type NM.DeviceState>, state-flags=<flags NM_ACTIVATION_STATE_FLAG_IS_MASTER | NM_ACTIVATION_STATE_FLAG_LAYER2_READY of type NM.ActivationStateFlags>                                                   
2020-04-09 10:26:45,831 root         DEBUG    NM action queue exhausted, quiting mainloop
2020-04-09 10:26:51,317 root         DEBUG    Checkpoint /org/freedesktop/NetworkManager/Checkpoint/1 rollback executed: dbus.Dictionary({dbus.String('/org/freedesktop/NetworkManager/Devices/10'): dbus.UInt32(0)
, dbus.String('/org/freedesktop/NetworkManager/Devices/3'): dbus.UInt32(0), dbus.String('/org/freedesktop/NetworkManager/Devices/2'): dbus.UInt32(0), dbus.String('/org/freedesktop/NetworkManager/Devices/8'): dbu
s.UInt32(0), dbus.String('/org/freedesktop/NetworkManager/Devices/5'): dbus.UInt32(0), dbus.String('/org/freedesktop/NetworkManager/Devices/6'): dbus.UInt32(0), dbus.String('/org/freedesktop/NetworkManager/Devic
es/7'): dbus.UInt32(0), dbus.String('/org/freedesktop/NetworkManager/Devices/1'): dbus.UInt32(0), dbus.String('/org/freedesktop/NetworkManager/Devices/9'): dbus.UInt32(0), dbus.String('/org/freedesktop/NetworkMa
nager/Devices/4'): dbus.UInt32(0)}, signature=dbus.Signature('su'))                                                                                                                                               
Traceback (most recent call last):
  File "/usr/bin/nmstatectl", line 11, in <module>
    load_entry_point('nmstate==0.2.9', 'console_scripts', 'nmstatectl')()
  File "/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py", line 59, in main
    return args.func(args)
  File "/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py", line 221, in apply
    return apply_state(statedata, args.verify, args.commit, args.timeout)
  File "/usr/lib/python3.6/site-packages/nmstatectl/nmstatectl.py", line 241, in apply_state
    rollback_timeout=timeout,
  File "/usr/lib/python3.6/site-packages/libnmstate/deprecation.py", line 40, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/libnmstate/nm/nmclient.py", line 96, in wrapped
    ret = func(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/libnmstate/netapplier.py", line 73, in apply
    state.State(desired_state), verify_change, commit, rollback_timeout
  File "/usr/lib/python3.6/site-packages/libnmstate/netapplier.py", line 175, in _apply_ifaces_state
    _verify_change(desired_state)
  File "/usr/lib/python3.6/site-packages/libnmstate/netapplier.py", line 221, in _verify_change
    verifiable_desired_state.verify_interfaces(current_state)
  File "/usr/lib/python3.6/site-packages/libnmstate/state.py", line 331, in verify_interfaces
    self._assert_interfaces_equal(other_state)
  File "/usr/lib/python3.6/site-packages/libnmstate/state.py", line 742, in _assert_interfaces_equal
    current_state.interfaces[ifname],
libnmstate.error.NmstateVerificationError: 
desired
=======
---
name: br1
type: linux-bridge
state: up
bridge:
  options:
    group-forward-mask: 0
    mac-ageing-time: 300
    multicast-snooping: true
    stp:
      enabled: false
      forward-delay: 15
      hello-time: 2
      max-age: 20
      priority: 32768
  port:
  - name: ens01
description: Linux bridge with the wrong port
ipv4:
  address: []
  auto-dns: true
  auto-gateway: true
  auto-routes: true
  dhcp: true
  enabled: true
ipv6:
  enabled: false
mac-address: 76:6D:F5:BE:84:8C
mtu: 1500

current
=======
---
name: br1
type: linux-bridge
state: up
bridge:
  options:
    group-forward-mask: 0
    mac-ageing-time: 300
    multicast-snooping: true
    stp:
      enabled: false
      forward-delay: 15
      hello-time: 2
      max-age: 20
      priority: 32768
  port: []
description: Linux bridge with the wrong port
ipv4:
  address: []
  auto-dns: true
  auto-gateway: true
  auto-routes: true
  dhcp: true
  enabled: true
ipv6:
  enabled: false
mac-address: 76:6D:F5:BE:84:8C
mtu: 1500

difference
==========
--- desired
+++ current
@@ -13,8 +13,7 @@
       hello-time: 2
       max-age: 20
       priority: 32768
-  port:
-  - name: ens01
+  port: []
 description: Linux bridge with the wrong port
 ipv4:
   address: []
```

Reply:
Cool cool. The yaml output is a bit shorter but I like the `difference` with this. 
The difference for the port is a good clue for this particular issue. In which case I could use that  instead of grepping for 'error'. Is that likely to be true for other troubleshooting issues (checking the diff)?

Reply:
With configuration issues, users would see this diff most of the time. So yes :)

Reply:
@phoracek as suggested, I've updated the command, the output, and the surrounding text. Let me know if there's any issues.